### PR TITLE
[ENH] single join fully implemented in numba

### LINKED ISSUE #1302: [ENH] implement full numba version of a single conditional_join
- currently the implementation is dependent on the non-numba implementation(generic_cond_join) - instead implement the core part completely in numba

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

- implement single conditional join fully in numba (for less than and greater than conditions)
- not equal operator implemented with numpy (no numba implementation)

**This PR resolves #1302 .**

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.md`.

3. [ ] Add a line to `CHANGELOG.md` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

# Automatic checks

There will be automatic checks run on the PR. These include:

- Building a preview of the docs on Netlify
- Automatically linting the code
- Making sure the code is documented
- Making sure that all tests are passed
- Making sure that code coverage doesn't go down.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by ericmjl** [comment]
🚀 Deployed on https://deploy-preview-1304--pyjanitor.netlify.app

File: tests/functions/test_conditional_join.py [hunk 1]
```diff
@@ -412,6 +412,69 @@ def test_single_condition_less_than_floats_keep_first(df, right):
...
+    )
+
+    assert_frame_equal(expected, actual)
+
+
+@pytest.mark.turtle
+@settings(deadline=None, max_examples=10)
+@given(df=conditional_df(), right=conditional_right())
+def test_single_condition_greater_than_floats_keep_last_numba(df, right):
+    """Test output for a single condition. "<"."""
```
Comment (ericmjl) [inline tests/functions/test_conditional_join.py]: @samukweku I took the liberty of asking ChatGPT for a better docstring on this test.

```python
@pytest.mark.turtle
@settings(deadline=None, max_examples=10)
@given(df=conditional_df(), right=conditional_right())
def test_single_condition_greater_than_floats_keep_last_numba(df, right):
    """
    Test the functionality of conditional_join with a single 'greater than' condition on floating-point data, while keeping the last match using Numba.

    This test sorts and filters dataframes 'df' and 'right' by columns 'B' and 'Numeric' respectively, removing NaN values. It then performs a backward merge_asof operation on these sorted dataframes. The expected outcome is a dataframe where each row from 'df' is merged with the last row from 'right' where 'Numeric' is greater than 'B'.

    The actual outcome is produced by the conditional_join method with a 'greater than' condition, left join type, sorted by appearance, keeping the last match, and utilizing Numba for performance optimization. The test asserts that the actual dataframe matches the expected dataframe, ensuring correct functionality of the conditional_join under these specific parameters.
    """
    # Test implementation continues...
```

Is it accurate? If so, I might begin writing a template for testing! Also, if it is accurate, could you update the docstrings for these two tests please? (I will get them generated for the other one.)

File: tests/functions/test_conditional_join.py [hunk 2]
```diff
@@ -412,6 +412,69 @@ def test_single_condition_less_than_floats_keep_first(df, right):
     assert_frame_equal(expected, actual)
 
 
+@pytest.mark.turtle
+@settings(deadline=None, max_examples=10)
+@given(df=conditional_df(), right=conditional_right())
+def test_single_condition_greater_than_floats_keep_last(df, right):
+    """Test output for a single condition. "<"."""
```
Comment (ericmjl) [inline tests/functions/test_conditional_join.py:416]:
```python
@pytest.mark.turtle
@settings(deadline=None, max_examples=10)
@given(df=conditional_df(), right=conditional_right())
def test_single_condition_greater_than_floats_keep_last_numba(df, right):
    """
    Test the functionality of conditional_join with a single 'greater than' condition on floating-point data, while keeping the last match using Numba.

    This test sorts and filters dataframes 'df' and 'right' by columns 'B' and 'Numeric' respectively, removing NaN values. It then performs a backward merge_asof operation on these sorted dataframes. The expected outcome is a dataframe where each row from 'df' is merged with the last row from 'right' where 'Numeric' is greater than 'B'.

    The actual outcome is produced by the conditional_join method with a 'greater than' condition, left join type, sorted by appearance, keeping the last match, without utilizing Numba for performance optimization. The test asserts that the actual dataframe matches the expected dataframe, ensuring correct functionality of the conditional_join under these specific parameters.
    """
    # Test implementation continues...
```

**Review by ericmjl — APPROVED** [review]
Pre-approving pending docstrings!

File: tests/functions/test_conditional_join.py (hunk 2 above)
Comment (samukweku) [reply tests/functions/test_conditional_join.py:416]: feels like a lot of docstring info for tests, no?

Comment (ericmjl) [reply tests/functions/test_conditional_join.py:416]: You have a good point, actually. I'm still a bit conflicted on whether to be verbose on test docstrings.

**Comment by ericmjl** [comment]
@samukweku are we good to merge? I think we should, please let me know.

**Comment by samukweku** [comment]
@ericmjl yes it is ok to merge

**Comment by ericmjl** [comment]
Thank you very much, @samukweku!

**Pull request merged**